### PR TITLE
[RISCV][docs] GP Relaxation and Small Data Limit

### PR DESCRIPTION
**Pull request by lenary**
As discussed in this week's RISC-V sync-up, we said we would add documentation about these options, and how they work.

---

I belive this is an accurate explanation, even though it is done in
reverse order (describing the linker related behaviour before the
compiler related behaviour).

Some discussion from before points to gp-relaxation + PIE being possible (if not implemented) - https://github.com/riscv-non-isa/riscv-elf-psabi-doc/issues/165#issuecomment-747158702 - so I have just documented the executable vs shared library difference, not anything about position-independence. I can certainly add to the description about position independence if needed.

Given this documentation was so target-specific, I thought it should go in one place on the RISC-V page rather than spread around LLD's and Clang's docs. I haven't yet done backreferences from Clang's and LLD's docs to this longer explanation, which I think would be useful.

Clarifications/Corrections welcome, I probably have missed something important.

### CODE REVIEW

**Comment by lenary** [comment]
Oh also, I have not documented what the default actually is for this, because I'm not sure it is agreed. I've also not documented non-llvm tools/libraries.

File: llvm/docs/RISCVUsage.rst [hunk 1]
```diff
@@ -431,3 +431,26 @@ line.  This currently applies to the following extensions:
...
+
+Some of the RISC-V psABIs reserve ``gp`` (``x3``) for use as a "Global Pointer", to make generating data addresses more efficient.
+
+To use this functionality, you need to:
+* not be using the ``gp`` register for any other uses -- some platforms use it for other things;
+* compile your objects with Clang's ``-mrelax`` option, to enable relaxation annotations on relocatable objects; and
+* be compiling for an executable (not a shared library); and
+* use LLD's ``--relax-gp`` option.
+
+LLD will relax (rewrite) any code sequences that materialize an address within 2048 bytes of this ``__global_pointer$`` (which will be defined if it does not already exist) to instead generate the address using ``gp`` and the correct (signed) 12-bit immediate. This usually saves at least one instruction compared to materialising a full 32-bit address value.
```
Comment (michaelmaitland) [inline llvm/docs/RISCVUsage.rst:446]: > this ``__global_pointer$``

I think this this the first time you mention `__global_pointer$` but the word `this` assumes that you had mentioned it before. Are you missing the introduction of this variable in the explanation?

**Review by michaelmaitland — COMMENTED** [review]
Thanks for this change. I would love for us to clear things up on PIE/PIC behavior in another PR at some point.

File: llvm/docs/RISCVUsage.rst [hunk 2]
```diff
@@ -431,3 +431,26 @@ line.  This currently applies to the following extensions:
...
+Global Pointer (GP) Relaxation and the Small Data Limit
+=======================================================
+
+Some of the RISC-V psABIs reserve ``gp`` (``x3``) for use as a "Global Pointer", to make generating data addresses more efficient.
+
+To use this functionality, you need to:
+* not be using the ``gp`` register for any other uses -- some platforms use it for other things;
+* compile your objects with Clang's ``-mrelax`` option, to enable relaxation annotations on relocatable objects; and
+* be compiling for an executable (not a shared library); and
+* use LLD's ``--relax-gp`` option.
```
Comment (topperc) [inline llvm/docs/RISCVUsage.rst:444]: lld only supports GP relaxation for medlow.

File: llvm/docs/RISCVUsage.rst (hunk 1 above)
Comment (lenary) [reply llvm/docs/RISCVUsage.rst:446]: I rewrote this several times, and i think the "this" just got left in. I think the sentence is clear enough without it, but I've also updated some later references to `__global_pointer$` with clarifications as well.

File: llvm/docs/RISCVUsage.rst (hunk 2 above)
Comment (lenary) [reply llvm/docs/RISCVUsage.rst:444]: Noted

File: llvm/docs/RISCVUsage.rst [hunk 3]
```diff
@@ -431,3 +431,29 @@ line.  This currently applies to the following extensions:
...
+
+There can only be one ``gp`` value in a process (as ``gp`` is not changed when calling into a function in a shared library), so the symbol is is only defined and this relaxation is only done for executables, and not for shared libraries. The linker expects executable startup code to put the value of ``__global_pointer$`` (from the executable) into ``gp`` before any user code is run.
+
+Arguably, the most efficient use for this addressing mode is for smaller global variables, as larger global variables are likely to need many more loads or stores when they are being accessed anyway.
+
+Therefore the compiler can place smaller global variables into sections with with names starting ``.sdata`` or ``.sbss`` (matching sections with names starting ``.data`` and ``.bss`` respectively). LLD knows to define the ``global_pointer$`` symbol close to these sections, and to lay these sections out adjacent to the ``.data`` section.
```
Comment (MaskRay) [inline llvm/docs/RISCVUsage.rst:453]: starting with?

File: llvm/docs/RISCVUsage.rst [hunk 4]
```diff
@@ -431,3 +431,29 @@ line.  This currently applies to the following extensions:
...
+* Compile for a position-dependent static executable (not a shared library, and ``-fno-PIC`` / ``-fno-pic`` / ``-fno-pie``); and
+* Use LLD's ``--relax-gp`` option.
+
+LLD will relax (rewrite) any code sequences that materialize an address within 2048 bytes of ``__global_pointer$`` (which will be defined if it is used and does not already exist) to instead generate the address using ``gp`` and the correct (signed) 12-bit immediate. This usually saves at least one instruction compared to materialising a full 32-bit address value.
+
+There can only be one ``gp`` value in a process (as ``gp`` is not changed when calling into a function in a shared library), so the symbol is is only defined and this relaxation is only done for executables, and not for shared libraries. The linker expects executable startup code to put the value of ``__global_pointer$`` (from the executable) into ``gp`` before any user code is run.
+
+Arguably, the most efficient use for this addressing mode is for smaller global variables, as larger global variables are likely to need many more loads or stores when they are being accessed anyway.
```
Comment (MaskRay) [inline llvm/docs/RISCVUsage.rst:451]: likely need many more

**Review by MaskRay — APPROVED** [review]

File: llvm/docs/RISCVUsage.rst (hunk 3 above)
Comment (lenary) [reply llvm/docs/RISCVUsage.rst:453]: ack.

File: llvm/docs/RISCVUsage.rst (hunk 4 above)
Comment (lenary) [reply llvm/docs/RISCVUsage.rst]: done.

**Review by michaelmaitland — APPROVED** [review]
LGTM

File: llvm/docs/RISCVUsage.rst [hunk 5]
```diff
@@ -431,3 +431,29 @@ line.  This currently applies to the following extensions:
...
 * ``Zvksh``
 * ``Zvkt``
+
+Global Pointer (GP) Relaxation and the Small Data Limit
+=======================================================
+
+Some of the RISC-V psABI variants reserve ``gp`` (``x3``) for use as a "Global Pointer", to make generating data addresses more efficient.
+
+To use this functionality, you need to be doing all of the following:
+* Use the ``medlow`` (aka ``small``) code model;
```
Comment (topperc) [inline llvm/docs/RISCVUsage.rst:441]: medany is supported with the bfd linker just not with lld.

File: llvm/docs/RISCVUsage.rst [hunk 6]
```diff
@@ -431,3 +431,29 @@ line.  This currently applies to the following extensions:
...
+* Not use the ``gp`` register for any other uses (some platforms use it for the shadow stack and others as a temporary -- as denoted by the ``Tag_RISCV_x3_reg_usage`` build attribute);
+* Compile your objects with Clang's ``-mrelax`` option, to enable relaxation annotations on relocatable objects;
```
Comment (topperc) [inline llvm/docs/RISCVUsage.rst:443]: `-mrelax` is the default isn't it? Should we mention that?

File: llvm/docs/RISCVUsage.rst (hunk 5 above)
Comment (lenary) [reply llvm/docs/RISCVUsage.rst:441]: I'm keen we just document LLVM here (LLD and Clang).

Comment (topperc) [inline llvm/docs/RISCVUsage.rst:441]: This isn't rendering as a bullet list https://llvm.org/docs/RISCVUsage.html#global-pointer-gp-relaxation-and-the-small-data-limit

Comment (lenary) [reply llvm/docs/RISCVUsage.rst:441]: Should be fixed on `main` now

**Pull request merged**